### DISCUSSION TITLE
Coupled thermo elasticity

### DISCUSSION THREAD

**Discussion by pcarruscag** [su2code/SU2#2402] [Ideas]
Continuation of #2398 and #2399 for @Vaish-W to work on.

The idea is to instantiate the heat solver alongside the FEA solver and then access the temperature at the nodes when we compute stresses, stiffness matrix, etc.

The instantiation of solvers happens in CSolverFactory::CreateSolverContainer, so look for places with `solver[FEA_SOL]` and add the creation of a heat solver conditional on some option, we can re-use WEAKLY_COUPLED_HEAT_EQUATION.

This should automatically create a SingleGridIntegration class for the heat problem together with the StructuralIntegration for the normal FEA problem.

Now we need to modify the CFEAIteration class to also integrate and advance the heat problem. This is in CFEAIteration::Iterate, before the `/*--- FEA equations ---*/` section we can add a call to the SingleGrid_Iteration of the heat integration class, see CFluidIteration.cpp around line 117 for what you need to do.

At this point we should have temperatures being computed by the heat solver and we can start using them in the FEA solver.
We need to do this in all functions that compute stresses Compute_StiffMatrix, Compute_StiffMatrix_NodalStressRes, Compute_NodalStressRes, and Compute_NodalStress.
Where these functions do `element->SetRef_Coord` we need to also `SetTemperature` (if it is available). These functions do not have access to heat solver at the moment, but CFEASolver::Preprocessing sees the entire solver container, so here we can store a pointer to the heat_solver or it's solution.

The last thing to do is to change CElasticityOutput to also output residual information and solution for the heat solver. You can look at CHeatOutput.cpp to see what is available in the heat solver.

Let me know if this makes sense to you and open a draft PR once you have some code so I can comment early on.
We also have weekly developer meetings we can use to talk about the project, the schedule and instructions are in our Slack channel.

**Comment by Vaish-W** [comment]
Thank you @pcarruscag for your detailed feedback on the pull request. I will surely work on this soon!!

**Comment by Vaish-W** [comment]
Hello,
I did first two steps as you suggested in PR #2404 but now stuck at preprocessing update. 

I have:
1. Introduced a CHeatSolver* heat_solver member to the CFEASolver class.
2. initialized it to nullptr in the constructor and deallocated it in the destructor.

Now moving to Preprocessor:
What would be approprate logic to preprocessing function which includes initializing the heat solver when it is enabled from config?

**Reply by pcarruscag** [reply]
Hi,
Make it just a CSolver* to make it simpler, then in preprocessing of the FEASolver you set this pointer to solver to the heat solver that was already created
heat_solver = solver_container[...][HEAT_SOL];
You don't need to deallocate because this is just a pointer to a solver that is managed (created and destroyed) outside of the FEA solver.

**Comment by Vaish-W** [comment]
Hey, how can I test the code we completed other than regression and unit tests? CI pipeline passed but should I prepare a problem for thermal stress simulation of a simple beam or plate?

**Reply by pcarruscag** [reply]
Yes, I would say the first verification step should be with a setup for the heat solver that gives an isothermal solution and comparing this with just setting the free stream temperature.
Then I would check an unloaded beam with linear temperature distribution, which has an easy analytical solution for the deformation due to thermal expansion.

**Reply by Vaish-W** [reply]
Can I use cantilever beam config_linear.cfg tutorial setup for this? And do these modifications:
Enable WEAKLY_COUPLED_HEAT_EQUATION.
Set all boundaries to isothermal (e.g., MARKER_ISOTHERMAL = (clamped, 300.0)).
Use FREESTREAM_TEMPERATURE = 300.0 for comparison

is it correct to do so?

**Reply by pcarruscag** [reply]
Yes, but in the case with weakly coupled heat set the free stream temperature to a different value from the isothermal value to make sure it's not used by mistake.

**Reply by Vaish-W** [reply]
[thermoelasticity.txt](https://github.com/user-attachments/files/18511543/thermoelasticity.txt)
I am getting following output where it gives 0 von mises stress for this cfg. 


```
-------------------- Solver Preprocessing ( Zone 0 ) --------------------
Filename: default_element_properties.dat.
There is no element-based properties file.
The structural domain has uniform properties.
Initialize Jacobian structure (Non-Linear Elasticity).
Initialize Jacobian structure (heat equation) MG level: 0.

------------------- Numerics Preprocessing ( Zone 0 ) -------------------

----------------- Integration Preprocessing ( Zone 0 ) ------------------

------------------- Iteration Preprocessing ( Zone 0 ) ------------------
FEM iteration.

------------------------------ Begin Solver -----------------------------

Simulation Run using the Single-zone Driver
CSysSolve::ConjugateGradient(): system solved by initial guess.
CSysSolve::ConjugateGradient(): system solved by initial guess.
+----------------------------------------------------------------+
|  Inner_Iter|  rms[DispX]|  rms[DispY]|      rms[T]|    VonMises|
+----------------------------------------------------------------+
|           0|  -32.000000|  -32.000000|  -32.000000|  0.0000e+00|

----------------------------- Solver Exit -------------------------------
All convergence criteria satisfied.
+-----------------------------------------------------------------------+
|      Convergence Field     |     Value    |   Criterion  |  Converged |
+-----------------------------------------------------------------------+
|                  rms[DispX]|           -32|          < -8|         Yes|
+-----------------------------------------------------------------------+
-------------------------------------------------------------------------
+-----------------------------------------------------------------------+
|        File Writing Summary       |              Filename             |
+-----------------------------------------------------------------------+
|SU2 binary restart                 |restart_linear_elastic.dat         |
|Paraview                           |linear_elastic.vtu                 |
+-----------------------------------------------------------------------+

------------------------- Exit Success (SU2_CFD) ------------------------
```



I have applied boundary condition for thermal as well as fea. What mistake did I make in cfg or could that be related to solver problem?

**Reply by Vaish-W** [reply]
If I remove left and clamped boundary condition from MARKER_ISOTHERMAL then I am getting stress as 1.1541e+06. Is that because fea and thermal boundary conditions are conflicting with each other and hence not accepting both condition for one marker?

+----------------------------------------------------------------+
|  Inner_Iter|  rms[DispX]|  rms[DispY]|      rms[T]|    VonMises|
+----------------------------------------------------------------+
|           0|   -9.644128|   -9.726810|  -32.000000|  1.1541e+06|

![image](https://github.com/user-attachments/assets/d8a13071-6229-4c01-8553-98c4d02064c4)

**Comment by pcarruscag** [comment]
Please reopen the pull request with the new branch

**Reply by Vaish-W** [reply]
Hi @pcarruscag, I have opened a pull request with new branch. Can you please tell me further steps?
Thank you!

**Reply by pcarruscag** [reply]
I think it is possible the marker definitions are overriding each other as you mentioned. So now you have to do some debugging to see if that's what is happening and if there is a way of fixing it.
We usually use config->GetMarker_All_KindBC(iMarker) to check the type of a marker, see for example CStructuralIntegration::Space_Integration_FEM and CIntegration::Space_Integration.
This can only return one value, so it's a problem for what we are trying to do...

We can try to address it by allowing CConfig to have different versions of Marker_All_KindBC internally which we switch when we call SetGlobalParam for example, config[val_iZone]->SetGlobalParam(MAIN_SOLVER::FEM_ELASTICITY, RUNTIME_FEA_SYS); would activate the FEA marker types, etc.

In the short term, and for testing purposes, you can try duplicating the marker definition inside the SU2 mesh file, and then use one copy for heat boundary conditions and another for FEA (this may break other things but it's worth a try).

**Reply by Vaish-W** [reply]
Hi @pcarruscag,
I tested with duplicate marker definition and it is working with that.
```
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% SU2 configuration file                                                       %
% Case description: Linear Thermo-Elasticity in Vertical Cantilever            %
% Author: Vaishnavi Wani                                                       %
% Institution: University of Stuttgart                                         %
% Date: 22-01-2025                                                             %
% File Version: 8.1.0 "Harrier"                                                %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%
% SOLVER TYPE
%%%%%%%%%%%%%%%%%%%%%%%   
          
SOLVER = ELASTICITY
WEAKLY_COUPLED_HEAT_EQUATION = YES
%%%%%%%%%%%%%%%%%%%%%%%
% STRUCTURAL PROPERTIES
%%%%%%%%%%%%%%%%%%%%%%%
GEOMETRIC_CONDITIONS = SMALL_DEFORMATIONS
MATERIAL_MODEL = LINEAR_ELASTIC
ELASTICITY_MODULUS = 5.0E9
POISSON_RATIO = 0.35
FORMULATION_ELASTICITY_2D = PLANE_STRESS
%%%%%%%%%%%%%%%%%%%%%%%
% INPUT
%%%%%%%%%%%%%%%%%%%%%%%
MESH_FORMAT = SU2
MESH_FILENAME = mesh_cantilever.su2
%%%%%%%%%%%%%%%%%%%%%%%
% BOUNDARY CONDITIONS
%%%%%%%%%%%%%%%%%%%%%%%
MARKER_CLAMPED = ( clamped )
MARKER_LOAD = ( left, 2.0, 1000, 1, 0, 0)
MARKER_PRESSURE = ( upper, 0, right, 0) 
MARKER_ISOTHERMAL = ( left_thermal, 300.0, right_thermal, 300.0 )
%%%%%%%%%%%%%%%%%%%%%%%
% FREESTREAM PROPERTIES
%%%%%%%%%%%%%%%%%%%%%%%
FREESTREAM_TEMPERATURE = 310.0
%%%%%%%%%%%%%%%%%%%%%%%
% LINEAR SOLVER
%%%%%%%%%%%%%%%%%%%%%%%
LINEAR_SOLVER = CONJUGATE_GRADIENT
LINEAR_SOLVER_PREC = ILU
LINEAR_SOLVER_ERROR = 1E-8
LINEAR_SOLVER_ITER = 1000
%%%%%%%%%%%%%%%%%%%%%%%
% OUTPUT
%%%%%%%%%%%%%%%%%%%%%%%
SCREEN_OUTPUT = (INNER_ITER, RMS_DISP_X, RMS_DISP_Y, RMS_TEMPERATURE, VMS)
TABULAR_FORMAT = CSV
OUTPUT_FILES =(RESTART, PARAVIEW)
CONV_FILENAME = history
VOLUME_FILENAME = linear_elastic
RESTART_FILENAME = restart_linear_elastic
```
OUTPUT:

```
------------------------------ Begin Solver -----------------------------

Simulation Run using the Single-zone Driver
CSysSolve::ConjugateGradient(): system solved by initial guess.
+----------------------------------------------------------------+
|  Inner_Iter|  rms[DispX]|  rms[DispY]|      rms[T]|    VonMises|
+----------------------------------------------------------------+
|           0|   -9.343098|   -9.425780|  -32.000000|  2.3082e+06|

----------------------------- Solver Exit -------------------------------
All convergence criteria satisfied.
+-----------------------------------------------------------------------+
|      Convergence Field     |     Value    |   Criterion  |  Converged |
+-----------------------------------------------------------------------+
|                  rms[DispX]|       -9.3431|          < -8|         Yes|
+-----------------------------------------------------------------------+
------------------------- Exit Success (SU2_CFD) ------------------------
```

![displacement](https://github.com/user-attachments/assets/9929399e-77cf-4df8-b425-6c986dbc9173)
![temp](https://github.com/user-attachments/assets/6b3e0350-1d06-40be-95c4-4fccdd98ff58)

**Reply by pcarruscag** [reply]
Nice, so now you can try some of the verification cases we talked about, and then we can work on a solution to avoid duplicating the markers.

**Reply by Vaish-W** [reply]
> Then I would check an unloaded beam with linear temperature distribution, which has an easy analytical solution for the deformation due to thermal expansion.

To do this, do I need to consider any extra parameters? Or just removing ```MARKER_LOAD``` and  ```MARKER_PRESSURE``` along with application of linear temperature distribution using ```MARKER_ISOTHERMAL = ( left_thermal, 300.0, right_thermal, 400.0 )``` would work?

**Reply by Vaish-W** [reply]
```
% SOLVER TYPE
SOLVER = ELASTICITY
WEAKLY_COUPLED_HEAT_EQUATION = YES
%%%%%%%%%%%%%%%%%%%%%%%
% STRUCTURAL PROPERTIES
%%%%%%%%%%%%%%%%%%%%%%%
GEOMETRIC_CONDITIONS = SMALL_DEFORMATIONS
MATERIAL_MODEL = LINEAR_ELASTIC
ELASTICITY_MODULUS = 5.0E9
POISSON_RATIO = 0.35
MATERIAL_THERMAL_EXPANSION_COEFF = 1.0E-5
FORMULATION_ELASTICITY_2D = PLANE_STRESS
MATERIAL_REFERENCE_TEMPERATURE = 300.0
%%%%%%%%%%%%%%%%%%%%%%%
% INPUT
%%%%%%%%%%%%%%%%%%%%%%%
MESH_FORMAT = SU2
MESH_FILENAME = mesh_cantilever.su2
%%%%%%%%%%%%%%%%%%%%%%%
% BOUNDARY CONDITIONS
%%%%%%%%%%%%%%%%%%%%%%%
MARKER_CLAMPED = ( clamped )
%MARKER_LOAD = ( left, 2.0, 1000, 1, 0, 0)
%MARKER_PRESSURE = (right, 0, left, 0) 
MARKER_ISOTHERMAL = (left_thermal, 350, right_thermal, 300, right, 0, left, 0, upper, 0)
%%%%%%%%%%%%%%%%%%%%%%%
% FREESTREAM PROPERTIES
%%%%%%%%%%%%%%%%%%%%%%%
FREESTREAM_TEMPERATURE = 310.0
%%%%%%%%%%%%%%%%%%%%%%%
% LINEAR SOLVER
%%%%%%%%%%%%%%%%%%%%%%%
LINEAR_SOLVER = CONJUGATE_GRADIENT
LINEAR_SOLVER_PREC = ILU
LINEAR_SOLVER_ERROR = 1E-8
LINEAR_SOLVER_ITER = 1000
%%%%%%%%%%%%%%%%%%%%%%%
% OUTPUT
%%%%%%%%%%%%%%%%%%%%%%%
SCREEN_OUTPUT = (INNER_ITER, RMS_DISP_X, RMS_DISP_Y, RMS_TEMPERATURE, VMS)
TABULAR_FORMAT = CSV
OUTPUT_FILES =(RESTART, PARAVIEW)
CONV_FILENAME = history
VOLUME_FILENAME = thermal_expansion
RESTART_FILENAME = restart_linear_elastic
```

I am getting no convergence with this code and no temperature variation from 300 to 400 degrees.

```
Simulation Run using the Single-zone Driver
CSysSolve::ConjugateGradient(): system solved by initial guess.
+----------------------------------------------------------------+
|  Inner_Iter|  rms[DispX]|  rms[DispY]|      rms[T]|    VonMises|
+----------------------------------------------------------------+
|           0|   -5.183450|   -5.282727|  -32.000000|  2.1779e+07|

----------------------------- Solver Exit -------------------------------
All convergence criteria satisfied.
+-----------------------------------------------------------------------+
|      Convergence Field     |     Value    |   Criterion  |  Converged |
+-----------------------------------------------------------------------+
|                  rms[DispX]|      -5.18345|          < -8|          No|
+-----------------------------------------------------------------------+
```

**Reply by pcarruscag** [reply]
Don't remove the markers, just make the load 0.

**Reply by Vaish-W** [reply]
Still getting same output with no convergence

**Reply by pcarruscag** [reply]
Well, this is what it's like to develop CFD codes :) you just have to debug things...

**Reply by Vaish-W** [reply]
Yes, I was just trying to understand if my .cfg file contains any errors before debugging the code... But will figure it out surely :)

**Reply by pcarruscag** [reply]
Some.
`MARKER_ISOTHERMAL = (left_thermal, 350, right_thermal, 300, right, 0, left, 0, upper, 0)`
You should not put right, left, etc. as 0 temperature, they should be 0 heat flux.
